{
 "metadata": {
  "name": "Django Notebook"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from django.core.management import setup_environ\n",
      "from loq import settings\n",
      "setup_environ(settings)\n",
      "import os\n",
      "os.environ['DJANGO_SETTINGS_MODULE'] = 'loq.settings'\n",
      "%pwd\n",
      "from loq.models import Read_alignment, Interval, Library"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### mapping lib -- obtained from csvimport to library table\n",
      "a = Read_alignment.objects.all()\n",
      "for i in a:\n",
      "    b = Library.objects.get(library_id=i.lib)\n",
      "    i.library = b\n",
      "    i.save()\n",
      "    #print i.library\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### formula to get readcounts from normreads and total mapped reads from each library\n",
      "import decimal\n",
      "for i in a:\n",
      "    i.read_counts = round(decimal.Decimal((i.normReads / 1000000) * i.library.total_mapped)* i.genomic_hits,0)\n",
      "    i.save()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### mapping foreign keys of interval to read_alignments using chr start stop and strand to verify. \n",
      "### currently this is missing strand to ensure both strands get mapped, but only sense is used for analysis \n",
      "a = Interval.objects.all()\n",
      "b = Read_alignment.objects.all()\n",
      "for i in a:\n",
      "    for j in b:\n",
      "        if j.chr == i.chr and j.start >= i.start and j.stop <= i.stop:\n",
      "            j.intervalName = i\n",
      "            #print j.intervalName\n",
      "            j.save()\n",
      "    #i.NeatName = str(i.chr)+ \":\" + str(i.start+1) +\"-\"+ str(i.stop)\n",
      "    #print i.NeatName\n",
      "    #i.IntervalSerialNumber = i.id\n",
      "   "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### using abundance of reads mapped to intervals to find out if interval is either + or - \n",
      "### this is not required if interval.csv file has mapped_strand header with + or - \n",
      "from django.db.models import Count\n",
      "\n",
      "for i in Interval.objects.all():\n",
      "    a = Read_alignment.objects.filter(intervalName=i).values('strand').annotate(Count('strand'))\n",
      "    #print a\n",
      "    \n",
      "    try:\n",
      "        if len(a) == 1:\n",
      "            i.mapped_strand = a[0]['strand']\n",
      "        elif len(a) == 0:\n",
      "            pass\n",
      "        else:\n",
      "            if a[0]['strand__count'] > a[1]['strand__count']:\n",
      "                i.mapped_strand = a[0]['strand']\n",
      "            else: \n",
      "                i.mapped_strand = a[1]['strand']\n",
      "        i.save()\n",
      "    except:\n",
      "        print i.id\n",
      "    #else:\n",
      "        #print i.id\n",
      "        #i.mapped_strand = a[0]['strand']\n",
      "     \n",
      "#print a[0]['strand']\n",
      "#print a[0]['strand']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### to fill in the read_length column using the length of string method of the read sequence\n",
      "a = Interval.objects.get(id=240)\n",
      "print a\n",
      "b = Read_alignment.objects.filter(chr='chrXHet', start__gte=183977, stop__lte=184045)\n",
      "for i in b:\n",
      "    #i.intervalName = a\n",
      "    i.read_length= len(i.sequence)\n",
      "    i.save()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### Parses the chromosome / genome file for the interval sequence according to the coordinates and strand\n",
      "from Bio import SeqIO\n",
      "from Bio.Seq import Seq\n",
      "\n",
      "self = Interval.objects.get(id=1)\n",
      "read_list= Read_alignment.objects.filter(intervalName=a)\n",
      "string =\"\"\n",
      "for i in read_list:\n",
      "    reads = '%s\\n' % (i.sequence)   \n",
      "    spaces = int(self.stop)-int(i.stop)\n",
      "    if spaces >= 1:\n",
      "        string += spaces * ' ' + reads \n",
      "    else:\n",
      "        string += reads\n",
      "#print read_list[0]\n",
      "        #print string\n",
      "#print Interval._meta.get_all_field_names()\n",
      "\n",
      "stt = self.start \n",
      "stp = self.stop\n",
      "#still hardcoded! needs to be static*\n",
      "   \n",
      "for seq_record in SeqIO.parse('/home/lichenhao/djangosite/loq/static/fasta/dmel-2L-chromosome-r5.50.fasta','fasta'):\n",
      "    seq = seq_record.seq\n",
      "    if Read_alignment.objects.filter(intervalName=a)[0].strand == \"+\":\n",
      "       seqInterval = str(seq[stt:stp])\n",
      "    else:\n",
      "        seqInterval = str(seq[stt:stp].reverse_complement())\n",
      "print seqInterval"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### 1st version of generating the coordinates for plotting the curve. \n",
      "### this uses a hash table to store values and updating via addition\n",
      "self= Interval.objects.get(id=2)\n",
      "reads= Read_alignment.objects.filter(intervalName=self).order_by('lib')\n",
      "table = {}\n",
      "list = []\n",
      "for i in range(self.start-1,self.stop+1):            \n",
      "    table[i]=int(0)\n",
      "for i in reads:\n",
      "    row = [i.start, i.stop, i.normReads]\n",
      "    for j in range(row[0],row[1]):\n",
      "        try:\n",
      "            if j in table:\n",
      "                table[j] = table[j] + row[2]\n",
      "        except:\n",
      "               print \"index not found\"\n",
      "for key, value in table.iteritems():\n",
      "    temp = [key,int(value)]\n",
      "    list.append(temp)\n",
      "    list= sorted(list)\n",
      "for i in list:\n",
      "    i[0]=str(i[0])\n",
      "print list\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### Used to generate the ucsc link for entries in the interval table \n",
      "a = Interval.objects.all()\n",
      "for i in a:\n",
      "    i.Link = \"http://genome.cbis.nus.edu.sg/cgi-bin/hgTracks?db=dm3&position=\" + str(i.NeatName)\n",
      "    i.save()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### finds the sum_normreads or sum_read_counts at each interval entry.\n",
      "from django.db.models import Sum\n",
      "a = Read_alignment.objects.all().values('intervalName').annotate(dsum=Sum('read_counts'))\n",
      "\n",
      "for i in a:\n",
      "    item = Interval.objects.get(id = i['intervalName'])\n",
      "    item.sum_read_counts= i['dsum']\n",
      "    print item.sum_read_counts\n",
      "    item.save()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import numpy as np\n",
      "from numpy.random import randn\n",
      "import pandas as pd\n",
      "from scipy import stats\n",
      "import matplotlib as mpl\n",
      "import matplotlib.pyplot as plt\n",
      "import seaborn as sns\n",
      "from matplotlib.pyplot import plot, hist"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### Code to update FK using previous FK\n",
      "self=Interval.objects.get(NeatName='chr3L:22766169-22767910')\n",
      "reads= Read_alignment.objects.filter(intervalName=self)\n",
      "for i in reads:\n",
      "    i.intervalName='chr3L:22766591-22766737'\n",
      "    i.save()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### code to store normReads using calculated formula\n",
      "aligns = Read_alignment.objects.filter(intervalName='chr3L:22766638-22766704')\n",
      "from srb.models import Library\n",
      "a = Library.objects.get(library_id='V064V0642')\n",
      "#print aligns\n",
      "for i in aligns:\n",
      "    #i.normReads = i.normRead()\n",
      "    i.library= a\n",
      "    i.save()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### Generates a pileup of either normreads or read_counts using pandas dataframe. similar to hash table\n",
      "### but allows for multiple libraries to be processed with overlaying plots.  \n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "%pylab inline\n",
      "self= Interval.objects.get(id=65)\n",
      "reads= Read_alignment.objects.filter(intervalName=self).order_by('lib')\n",
      "\n",
      "libs = ['D005', 'D006','D007','D008','D009', 'D010']\n",
      "cords = np.arange(self.start-1,self.stop+1)\n",
      "w = range(self.start-1,self.stop+1)\n",
      "w = np.array([w])\n",
      " \n",
      "df = pd.DataFrame(index=cords, columns=libs)\n",
      "df = df.fillna(0)\n",
      "\n",
      "for i in reads:\n",
      "    for j in range(i.start,i.stop):\n",
      "        df[i.lib][j] += i.normReads\n",
      "\n",
      "a= np.array(df)\n",
      "w = np.reshape(w,(len(cords),-1))\n",
      "b = np.concatenate((w,a),axis=1)\n",
      "b = np.flipud(b)\n",
      "b = b.tolist()\n",
      "#libs.insert(0,\"Coordinates\")\n",
      "#b.insert(0,libs)\n",
      "for i in b:\n",
      "    i[0] =str(i[0])\n",
      "\n",
      "#row_1=['Coordinates']\n",
      "#for i in libs:\n",
      "#    row_1.append(str(i))\n",
      "\n",
      "#b = b.insert(0,row_1)\n",
      "\n",
      "#len(a)\n",
      "#print df[0:50]\n",
      "#cord = np.transpose(cords)\n",
      "\n",
      "fig= plt.figure()\n",
      "ax = fig.add_subplot(111)\n",
      "ax.plot(df)\n",
      "#plt.ylim(0,3.0)\n",
      "plt.title('Reads pileup for ' +str(self.mirName))\n",
      "#plt.xlim(5508754,5508838)\n",
      "ax.xaxis.get_major_formatter().set_powerlimits((0, 5))\n",
      "plt.legend(libs)\n",
      "#df.plot()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### Deprecated code to calculate pileup\n",
      "self=Interval.objects.get(IntervalSerialNumber=1)\n",
      "reads= Read_alignment.objects.filter(library_id='V064V0642')\n",
      "cord = []\n",
      "data =[]\n",
      "index = []\n",
      "for i in reads:\n",
      "    index.append(i.id)\n",
      "    cord.append(i.start)\n",
      "    cord.append(i.stop)\n",
      "    ac =[int(i.start-self.start)*str(0),int(i.stop-i.start) *str(i.read_counts),int(self.stop-i.stop)*str(0)]\n",
      "    #print list(\"\".join(ac))\n",
      "    data.append(list(\"\".join(ac)))\n",
      "#for i in data: \n",
      "#    print i\n",
      "#cords = range(self.start,self.stop)\n",
      "#for i in data: \n",
      "#    i=map(int,i)\n",
      "#print data\n",
      "#    matrix\n",
      "#matrix= zip(matrix)\n",
      "#print matrix\n",
      "#print len(cords)\n",
      "#print len(index)\n",
      "#print np.array(matrix)\n",
      "#print np.array(data)\n",
      "#frame = pd.DataFrame(data,columns=cords, index=index).convert_objects(convert_numeric=True)\n",
      "#print frame[22766591]\n",
      "#pile= frame.values.sum(axis=1)\n",
      "#print pile\n",
      "#graph_list=[]\n",
      "#print reads\n",
      "#print str(data)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### 2nd version of plotting using a list to contain values. old and deprecated\n",
      "self=Interval.objects.get(NeatName='chr3L:22766591-22766737')\n",
      "reads= Read_alignment.objects.filter(intervalName=self)\n",
      "cord = []\n",
      "data =[]        \n",
      "index = []\n",
      "for i in reads:\n",
      "    index.append(i.id)\n",
      "    cord.append(i.start)\n",
      "    cord.append(i.stop)\n",
      "    ac =[int(i.start-self.start)*str(0),int(i.stop-i.start) *str(i.read_counts),int(self.stop-i.stop)*str(0)]\n",
      "    data.append(list(\"\".join(ac)))\n",
      "    nt=sorted(cord)\n",
      "    cords = range(self.start,self.stop)\n",
      "    matrix = [] \n",
      "for i in data:\n",
      "    i=map(int,i)\n",
      "    #print len(i)\n",
      "    matrix.append(i)\n",
      "frame = pd.DataFrame(matrix,columns=cords, index=index, dtype='int')\n",
      "pile= frame.values.sum(axis=0)\n",
      "g_list= []\n",
      "for i in range(0,len(cords)):\n",
      "    g_list.append([str(cords[i]),pile[i]])    \n",
      "g_list.insert(0,[\"Position(nt)\", \"Counts\"])\n",
      "g_list"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "### old script, non functional\n",
      "#from django.db.models import Count\n",
      "#from pandas import DataFrame\n",
      "#def toDict(x):\n",
      "#    return [item for item in x]\n",
      "#lib_table ={}\n",
      "\n",
      "#print libs\n",
      "#table = {}\n",
      "#frame ={}\n",
      "\n",
      "#self= Interval.objects.get(id=2)\n",
      "\n",
      "#reads= Read_alignment.objects.filter(intervalName=self).order_by('lib')\n",
      "#libraries = reads.values('lib').annotate(dcount=Count('lib'))\n",
      "\n",
      "\n",
      "#for i in range(self.start-1,self.stop+1):            \n",
      "#    table[i]=int(0)\n",
      "    \n",
      "#lib_dict = toDict(libraries)\n",
      "#libs = [i['lib'] for i in lib_dict]\n",
      "#libs = ['D005', 'D006','D007','D008','D009', 'D010']\n",
      "#frame ={i:table for i in libs}\n",
      "#print frame\n",
      "#for i in libs:\n",
      "#    data = Read_alignment.objects.filter(intervalName=self, lib=i)\n",
      "#print frame\n",
      "data = []\n",
      "for j in reads: \n",
      "    data.append([j.lib,j.start,j.stop,j.normReads])\n",
      "    \n",
      " \n",
      "\n",
      "for i in data:\n",
      "    for q in range(i[1],i[2]+1):\n",
      "        #currentReads = frame[1][q]\n",
      "        #print currentReads\n",
      "        #frame[i[0]][q] = i[3] + currentReads\n",
      "    #print frame['D005'] == frame['D010']    \n",
      "            # OR\n",
      "        frame[i[0]][q] += j.normReads\n",
      "            #frame[j.lib] =table\n",
      "            #table[q] += j.normReads\n",
      "                        \n",
      "#print j.lib +'\\t'+ str(frame[][857595]) + '\\t'+ str(frame['D005'][857595])  \n",
      "\n",
      "print frame['D005']       \n",
      "print '\\n\\n'\n",
      "print frame['D008']\n",
      "#print frame['D006']    \n",
      "#print frame['D008'] == frame['D005']\n",
      "\n",
      "#for k,v in lib_table.iteritems():\n",
      "#print lib_table\n",
      "#array = []\n",
      "#for keys, values in lib_table.iteritems():\n",
      "#    for k,v in values.iteritems():\n",
      "#        array.append([k,v])\n",
      "#    lib_table[keys] = sorted(array)\n",
      "    #print len(array)\n",
      "#    array =[]\n",
      "#left = DataFrame(lib_table)\n",
      "#print left\n",
      "        #list.append(temp)\n",
      "        #list= sorted(list)\n",
      "#for i in list:\n",
      "#    i[0]=str(i[0])\n",
      "#print list\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}